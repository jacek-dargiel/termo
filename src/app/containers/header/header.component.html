<header>
  <button
    class="refresh"
  >
    <div
      class="refresh__progress"
      [style.width.%]="progress | async"
    ></div>
    <span class="refresh__label">Odśwież</span>
  </button>
</header>