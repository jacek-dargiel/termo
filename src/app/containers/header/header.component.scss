@import '../../../styles/variables';

:host {
  position: absolute;
  top: 0;
  display: block;
  width: 100%;
  z-index: 2;
}

header {
  padding: 0 1.5rem;
  height: $headerHeight;
  // background: $colorPrimary;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: flex-end;

  @media (max-width: 1024px) {
    padding: 0 1rem;
  }
}

.refresh {
  background: lighten($colorPrimary, 10);
  box-shadow: 0 0 2rem 0.5rem rgba(darken($colorPrimary, 30), 0.5);
  border: 0;
  color: #fff;
  text-transform: uppercase;
  font-weight: 500;
  font-size: 1.2rem;
  padding: 0.75rem 4rem;
  position: relative;

  &:focus,
  &:active {
    outline: 0.2rem solid #eede25;
  }

  &__progress {
    position: absolute;
    width: 0;
    background-color: lighten($colorPrimary, 20);
    height: 100%;
    top: 0;
    left: 0;
    z-index: 1;
    transition: width 1s linear;

    &--refreshing {
      background-color: lighten($colorPrimary, 10);
      background-image: linear-gradient(-45deg, rgba(255, 255, 255, 0.35) 25%, rgba(255, 255, 255, 0) 25%, rgba(255, 255, 255, 0) 50%, rgba(255, 255, 255, 0.35) 50%, rgba(255, 255, 255, 0.35) 75%, rgba(255, 255, 255, 0) 75%, rgba(255, 255, 255, 0));
      background-size: 20px 20px;
      animation: candystripe 1s linear infinite;
      width: 100% !important;
      transition: width 0s linear;
    }
  }
  

  &__label {
    z-index: 2;
    position: relative;
  }
}

@keyframes candystripe {
  to {background-position: 20px 0;}
}
