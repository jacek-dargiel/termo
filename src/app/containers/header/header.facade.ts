import { Injectable } from '@angular/core';
import { RefreshSignalService } from '../../services/refresh-signal.service';
import { map, first } from 'rxjs/operators';
import { environment } from 'environments/environment';
import { Store } from '@ngrx/store';
import { State } from '../../state/reducers';
import { selectMeasurmentsLoading, selectAllLocations } from '../../state/selectors';
import { RefreshMeasurmentsStart } from '../../state/location/location.actions';

@Injectable({
  providedIn: 'root',
})
export class HeaderFacade {
  public progress = this.refreshSignal.signal.pipe(
    map(timeLeft => timeLeft * 1000),
    map(timeLeft => (environment.refreshTimeout - timeLeft) / environment.refreshTimeout),
  );
  public refreshing = this.store.select(selectMeasurmentsLoading);

  constructor (
    private refreshSignal: RefreshSignalService,
    private store: Store<State>
  ) {}

  refresh() {
    this.store.select(selectAllLocations).pipe(
      first(),
    )
      .subscribe(locations => this.store.dispatch(new RefreshMeasurmentsStart({ locations })));
  }
}
