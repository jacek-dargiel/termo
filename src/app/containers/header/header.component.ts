import { Component } from '@angular/core';
import { HeaderFacade } from './header.facade';

@Component({
    selector: 'termo-header',
    templateUrl: './header.component.html',
    styleUrls: ['./header.component.scss'],
    standalone: false
})
export class HeaderComponent {
  public progress$ = this.headerFacade.progress;
  public refreshing$ = this.headerFacade.refreshing;

  constructor(
    private headerFacade: HeaderFacade,
  ) { }

  public refresh () {
    this.headerFacade.refresh();
  }

}
