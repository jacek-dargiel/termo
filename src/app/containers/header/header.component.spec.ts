import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';

import { HeaderComponent } from './header.component';
import { HeaderFacade } from './header.facade';
import { ReplaySubject } from 'rxjs';
import { NO_ERRORS_SCHEMA } from '@angular/core';

class MockHeaderFacade {
  public progress = new ReplaySubject<number>();
  public refreshing = new ReplaySubject<boolean>();

  constructor() {
    this.progress.next(0);
    this.refreshing.next(false);
  }

  public refresh = jest.fn();
}

describe('HeaderComponent', () => {
  let component: HeaderComponent;
  let fixture: ComponentFixture<HeaderComponent>;
  let facade: MockHeaderFacade;

  beforeEach(waitForAsync(() => {
    TestBed.configureTestingModule({
      declarations: [ HeaderComponent ],
      providers: [
        { provide: HeaderFacade, useClass: MockHeaderFacade }
      ],
      schemas: [NO_ERRORS_SCHEMA],
    })
    .compileComponents();
    facade = TestBed.get(HeaderFacade);
  }));

  beforeEach(() => {
    fixture = TestBed.createComponent(HeaderComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();

  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should render initial state', () => {
    expect(fixture).toMatchSnapshot();
  });

  it('should refresh when "Refresh" button clicked', () => {
    component.refresh();
    expect(facade.refresh).toHaveBeenCalled();
  });

});
