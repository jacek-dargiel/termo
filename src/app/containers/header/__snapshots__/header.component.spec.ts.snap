// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HeaderComponent should render initial state 1`] = `
<termo-header
  headerFacade={[Function MockHeaderFacade]}
  progress$={[Function ReplaySubject]}
  refreshing$={[Function ReplaySubject]}
>
  <header>
    <termo-refresh-button>
      Odśwież
    </termo-refresh-button>
  </header>
</termo-header>
`;
